File: frontend/src/lib/polyline.ts
/**
 * Decoder for Google's Encoded Polyline Algorithm Format with the standard
 * precision of 5 decimal places. The encoding stores deltas of integer-scaled
 * lat/lon coordinates as variable-length, ASCII-printable chunks.
 *
 * Mirrors the encoder used on the backend (track.EncodePolyline) so a
 * server-encoded polyline round-trips through the wire and back to lat/lon
 * pairs without precision loss beyond ~1.1m at the equator.
 */

const POLYLINE_PRECISION = 1e5

export interface LatLon {
  lat: number
  lon: number
}

/**
 * Decodes an encoded polyline string into an array of WGS84 lat/lon points.
 * Returns an empty array for empty input.
 *
 * The string is consumed byte-by-byte: each value is read by accumulating
 * 5-bit chunks until a chunk with the high (0x20) bit cleared is seen, and
 * the resulting unsigned integer is then zig-zag decoded back to a signed
 * delta. The first delta is relative to (0, 0); subsequent deltas accumulate.
 */
export function decodePolyline(encoded: string): LatLon[] {
  const out: LatLon[] = []
  if (!encoded) return out

  let index = 0
  let lat = 0
  let lon = 0
  while (index < encoded.length) {
    const [dLat, nextLat] = readValue(encoded, index)
    if (nextLat < 0) break
    lat += dLat
    index = nextLat
    if (index >= encoded.length) break
    const [dLon, nextLon] = readValue(encoded, index)
    if (nextLon < 0) break
    lon += dLon
    index = nextLon
    out.push({ lat: lat / POLYLINE_PRECISION, lon: lon / POLYLINE_PRECISION })
  }
  return out
}

/**
 * Reads a single zig-zag encoded signed integer from the encoded polyline at
 * pos. Returns [value, nextPos] on success or [0, -1] on truncated input.
 */
function readValue(s: string, pos: number): [number, number] {
  let result = 0
  let shift = 0
  let p = pos
  while (true) {
    if (p >= s.length) return [0, -1]
    const b = s.charCodeAt(p) - 63
    p++
    result |= (b & 0x1f) << shift
    shift += 5
    if (b < 0x20) break
    if (shift > 30) return [0, -1]
  }
  // Zig-zag decode.
  const value = result & 1 ? ~(result >>> 1) : result >>> 1
  return [value, p]
}

File: internal/pkg/track/polyline.go
package track

import (
	"math"
	"strings"
)

// polylinePrecision is the coordinate precision used by EncodePolyline and
// DecodePolyline (precision 5, ~1.1m at the equator). This matches Google's
// original "Encoded Polyline Algorithm Format".
const polylinePrecision = 1e5

// EncodePolyline encodes a sequence of points using Google's polyline
// algorithm at precision 5. Only latitude and longitude are encoded;
// elevation and timestamps are ignored. Returns "" for an empty input.
func EncodePolyline(pts Points) string {
	if len(pts) == 0 {
		return ""
	}
	var b strings.Builder
	b.Grow(len(pts) * 8)

	prevLat := int32(0)
	prevLon := int32(0)
	for _, p := range pts {
		lat := int32(math.Round(p.Lat * polylinePrecision))
		lon := int32(math.Round(p.Lon * polylinePrecision))
		appendPolylineValue(&b, lat-prevLat)
		appendPolylineValue(&b, lon-prevLon)
		prevLat = lat
		prevLon = lon
	}
	return b.String()
}

// appendPolylineValue zig-zag encodes a signed int32 delta and writes the
// resulting variable-length quantity into b using the printable-ASCII
// encoding from the Google polyline format.
func appendPolylineValue(b *strings.Builder, v int32) {
	// Zig-zag transform: positive values map to even, negative to odd.
	uv := uint32(v) << 1
	if v < 0 {
		uv = ^uint32(v)<<1 | 1
	}
	for uv >= 0x20 {
		b.WriteByte(byte((0x20 | (uv & 0x1f)) + 63))
		uv >>= 5
	}
	b.WriteByte(byte(uv + 63))
}

// DecodePolyline decodes a string produced by EncodePolyline (precision 5)
// back into a sequence of points. Elevation is left at zero. Returns nil
// for an empty string. A malformed trailing fragment is silently dropped.
func DecodePolyline(s string) Points {
	if s == "" {
		return nil
	}
	var (
		out    Points
		i      int
		curLat int32
		curLon int32
	)
	for i < len(s) {
		dLat, n, ok := readPolylineValue(s, i)
		if !ok {
			break
		}
		i += n
		dLon, n, ok := readPolylineValue(s, i)
		if !ok {
			break
		}
		i += n
		curLat += dLat
		curLon += dLon
		out = append(out, Point{
			Lat: float64(curLat) / polylinePrecision,
			Lon: float64(curLon) / polylinePrecision,
		})
	}
	return out
}

// readPolylineValue reads a single zig-zag encoded value from s starting at
// pos and returns the decoded value, the number of bytes consumed, and a
// success flag. ok is false when the encoded value is truncated.
func readPolylineValue(s string, pos int) (int32, int, bool) {
	var (
		result uint32
		shift  uint
		n      int
	)
	for {
		if pos+n >= len(s) {
			return 0, 0, false
		}
		b := uint32(s[pos+n]) - 63
		n++
		result |= (b & 0x1f) << shift
		shift += 5
		if b < 0x20 {
			break
		}
	}
	if result&1 != 0 {
		return ^int32(result >> 1), n, true
	}
	return int32(result >> 1), n, true
}
